py3status/modules/google_calendar.py

```diff
@@ -0,0 +1,502 @@
+    2. OAuth 2.0 credentials for the Google Calendar api.
+
+    Follow Step 1 of the guide here to obtain your OAuth 2.0 credentials:
+    https://developers.google.com/google-apps/calendar/quickstart/python
+
+    Download the client_secret.json file which contains your client ID and
+    client secret. In your i3status config file, set configuration parameter
+    client_secret to the path to your client_secret.json file.
+
+    The first time you run the module, a browser window will open asking you
+    to authorize access to your calendar. After authorization is complete,
+    an access/refresh token will be saved to the path configured in
+    auth_token, and i3status will be restarted. This restart will
+    occur only once after the first time you successfully authorize.
+
+Examples:
+```
+# add color gradients for events and dates/times
+google_calendar {
+    thresholds = {
+        'event': [(1, '#d0e6ff'), (2, '#bbdaff'), (3, '#99c7ff'), (4, '#86bcff'),
+            (5, '#62a9ff'), (6, '#8c8cff'), (7, '#7979ff')],
+        'time': [(1, '#ffcece'), (2, '#ffbfbf'), (3, '#ff9f9f'), (4, '#ff7f7f'),
+            (5, '#ff5f5f'), (6, '#ff3f3f'), (7, '#ff1f1f')]
+    }
+}
+```
+
+@author Igor Grebenkov
+@license BSD
+
+SAMPLE OUTPUT
+[
+    {
+      'full_text':'Homer's Birthday (742 Evergreen Terrace) (1h 23m)',
+      'index':0,
+      'separator':False,
+      'separator_block_width':0,
+      'cached_until':1509008337.0
+   },
+   {
+      'full_text':'|',
+      'index':'sep',
+      'separator':False,
+      'separator_block_width':0
+   },
+   {
+      'full_text':'Doctor's Appointment',
+      'index':1,
+      'separator':False,
+      'separator_block_width':0
+   },
+   {
+      'full_text':'|',
+      'index':'sep',
+      'separator':False,
+      'separator_block_width':0
+   },
+   {
+      'full_text':'Lunch with John',
+      'index':2
+   }
+]
+"""
+
+import httplib2
+import os
+import datetime
+
+from apiclient import discovery
+from oauth2client import client
+from oauth2client import clientsecrets
+from oauth2client import tools
+from oauth2client.file import Storage
+from httplib2 import ServerNotFoundError
+from dateutil import parser
+from dateutil.tz import tzlocal
+
+
+SCOPES = 'https://www.googleapis.com/auth/calendar.readonly'
+APPLICATION_NAME = 'py3status google_calendar module'
+
+
+class Py3status:
+    """
+    """
+    auth_token = '~/.config/py3status/google_calendar.auth_token'
+    blacklist_events = []
+    button_open = 3
+    button_refresh = 2
+    button_toggle = 1
+    cache_timeout = 60
+    client_secret = '~/.config/py3status/google_calendar.client_secret'
+    force_lowercase = False
+    format = '{events}|\?color=event \u2687'
+    format_date = '%a %d-%m'
+    format_event = '[\?color=event {summary}][\?if=is_toggled  ({start_time}' +\
+        ' - {end_time}, {start_date})|[\?if=location  ({location})] {format_timer}]'
```

Comment:
(Q) Using `[ ({location})]` instead of `[\?if=location  ({location})]` does not work here?

Reply:
Uhh, I thought it wasn't working last night, but it is now. I dunno how the hell that happened. I'll change it.

Reply:
Oh, I remember why now. Without `\?if=location`, when `is_current = True` I was getting `invalid format` for an event with no location. `safe_format` would throw an IndexError.  Wrapping `{format_timer}` in brackets like so: `[{format_timer}]` fixed it.

Reply:
@tobes Can you look at this?
```
WARNING Invalid format `[\?color=event {summary}][\?if=is_toggled  ({start_time} - {end_time}, {start_date})|[({location})] {format_timer}]` (IndexError) google_calendar.py line 448.
```
If I give it a space, then it'll work okay.
```
... _time}, {start_date})| [({location})] {form... 
                          ^
```
Thanks. :+1: